# NGS2_project
Genomic evaluation through accounting for recognized variant locations is one of the aims of bioinformatics analysis. Calling of variants either during alignment or after it may represents as one of the challenging for improving speed and accuracy of sequence alignment. The previous known Burrowsâ€“Wheeler alignment (BWA) and variant identification using Genome Analysis ToolKit (GATK) and SAMtools are compared with Findmap and Findvar programs which are used during the alignment through the present study. Otherwise, due to troubleshooting, different programs may be used in the comparison such as Freebayes. Strategies using human colorectal tumor sample and a whole human reference chromosome 18 are tested. Aligning reads and counting reference and alternate alleles for DNA source, outputs potential new single nucleotide polymorphysim, insertion, and deletion alleles are carried out. Tests are assessed according to times for processing for BWA, GATK, and SAMtools on one side, and for the other program on the other in minutes. Alignment programs requiring total memory (in GB) for each program are also well-thought-out. Mapping of reads and accuracy of calling alleles for known variants are valued. Advantages of either both programs through processing rapidness, precise alignment, usefulness of data summaries, compact output, and number of steps are reported in which calling known variants more efficiently and accurately are reflected.
Key words: Sequence alignment, Variant calling, GATK, BWA, Findmap, Findvar, Freebayes
